Feat camera calibration

gliegard opened the pull request:
## Description
- OrientedImageCamera : Extension of a ThreeJs Camera adapted to computer vision.
- CameraCalibrationParser : Create a set of  OrientedImageCamera from a configuration file
- an example show a car with 5 cameras

## Motivation and Context
Long-term goal : Immersive view

## Screenshots
![camera_parser](https://user-images.githubusercontent.com/15127065/46739667-5f91ed00-cca1-11e8-8c9e-e5d54addbaae.jpg)

gliegard committed:
feat(OrientedImageCamera): Camera extension for oriented images

gliegard committed:
feat(parser): Camera calibration parser

zarov commented:
I feel this is the same example as the current `orientation_utils` examples, why not update those instead ?

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,80 @@
+ * @module CameraCalibrationParser
+ */
+
+THREE.Matrix4.prototype.setMatrix3 = function setMatrix3(m) {
```
What about renaming it to `setFromMatrix3` ? Did you also try to open discussion to add this method in THREE.js ? I didn't find anything like that on their repository, so I don't know if it's relevant to suggest it.

mbredif replied:
`setFromMatrix3` would imply that the whole Matrix4 is set, whereas this only updates the upper left Matrix3.
`setMatrix3` is somewhat more in line with `setPosition` which only updates the upper right Vector3.
I definitely support an upstream PR.

zarov replied:
Ok for `setMatrix3`

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,80 @@
+    return this;
+};
+
+// the json format encodes the following transformation:
```
This description should be below in the `parse` documentation, so we can read it in the generated documentation, or above in the `module` section.

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,80 @@
+    // p_world = position + transpose(rotation) * p_local
+    camera.position.fromArray(calibration.position);
+    // calibration.rotation is row-major but fromArray expects a column-major array, yielding the transposed matrix
+    var rotationInverse = new THREE.Matrix3().fromArray(calibration.rotation);
```
Why not alter the `calibration.rotation` array, to be a 16-length array, and then create directly a `Matrix4` array ? Or at least you should declare the `rotationInverse` Matrix3 outside this function.

mbredif replied:
A 4x4 rotation would have 7 constant entries (0s on the last lines and columns except a 1 on the lower right) which seems a bit redundant...
Ok for the memory optimization.

zarov replied:
Yeah I was kind of conflicted on this one, just move the `new` outside the method.

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,80 @@
+    const focal = new THREE.Vector2(proj[0], proj[4]);
+    const center = new THREE.Vector2(proj[2], proj[5]);
+    const skew = proj[1];
+    var camera = new OrientedImageCamera(size, focal, center, options.near, options.far, skew);
```
Don't use `var`.

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,80 @@
+
+    // parse distortion
+    if (calibration.distortion) {
+        camera.distortion = {
```
What is the purpose of the distortion property ? I find no other mention of it. Also it should be at least initialized in `OrientedImageCamera`.

gliegard replied:
It will be used in #866 https://github.com/iTowns/itowns/blob/6293c10d33328a3068ee23fb6dab35cee4aebf5b/src/Renderer/OrientedImageMaterial.js#L61 
It can be defined and used, or it can be undefined and not used.

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,80 @@
+    // parse distortion
+    if (calibration.distortion) {
+        camera.distortion = {
+            pps: new THREE.Vector2().fromArray(calibration.distortion.pps),
```
I don't know the usage you will have of this, but is it necessary to be Vectors ?

mbredif replied:
I think so : it will be passed as a `vec2` uniform.

zarov replied:
Oh yeah makes sense, thanks

zarov commented on src/Renderer/OrientedImageCamera.js:
```diff
@@ -0,0 +1,60 @@
+import * as THREE from 'three';
+
+class OrientedImageCamera extends THREE.PerspectiveCamera {
```
We need more information on the purpose of this new camera, please add some documentation.

mbredif replied:
I continued working on this camera after handing it to @gliegard and my version has a better documentation that you can adapt/reuse : https://github.com/mbredif/three-photogrammetric-camera/blob/master/src/cameras/PhotogrammetricCamera.js

gliegard replied:
ok, I've add documentation.
Thanks Mathieu.

zarov commented on src/Renderer/OrientedImageCamera.js:
```diff
@@ -0,0 +1,60 @@
+        this.center = center || size.clone().multiplyScalar(0.5);
+        this.skew = skew || 0;
+        Object.defineProperty(this, 'fov', {
+            get: () => Math.atan2(this.size.y, 2 * this.focal.y) * 360 / Math.PI,
```
This value should be computed somewhere else, not every time we ask for the fov.

mbredif replied:
This class makes `fov` a computed property depending on size.y and focal. An alternative would be to use setters for size and focal that update a cached fov value.

zarov replied:
Sounds good to leave as it is then, no problem

zarov reviewed (changes requested):
A few changes to make. Also you should add a unit test for the parser with a few cameras, and the `setMatrix3` method also.

mbredif reviewed (commented):
Looks good !
+1 for extending the existing example instead of creating a new one.

gliegard committed:
example(cameraCalibrationParser): Example showing 5 cameras helper on a car.
This example create 5 OrientedImageCamera,
reading a description file using the CameraCalibrationParser.

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,82 @@
+     * @param {string|JSON} json - the json content of the calibration file.
+     * @param {Object} options - Options controlling the parsing.
+     * @param {string} options.near - Near of the created cameras.
+     * @param {string} options.near - Far of the created cameras.
```
`options.far`

zarov commented on src/Parser/CameraCalibrationParser.js:
```diff
@@ -0,0 +1,82 @@
+}
+
+export default {
+    /**
```
Could you align the documentation for something similar to "clean" parser ? Like `GeojsonParser` or `VectorTileParser` for example.

zarov commented on src/Renderer/OrientedImageCamera.js:
```diff
@@ -0,0 +1,74 @@
+ */
+class OrientedImageCamera extends THREE.PerspectiveCamera {
+    /**
+     * @Constructor
```
`@constructor`

zarov commented on src/Renderer/OrientedImageCamera.js:
```diff
@@ -0,0 +1,74 @@
+        this.updateProjectionMatrix();
+    }
+
+    updateProjectionMatrix() {
```
Please comment why you are overriding this method.

zarov reviewed (commented):
A few things to change, only documentation. Thanks for adding unit tests !

zarov reviewed (approved):
Thanks for these changes !
